src/runtime/interop.cs

```diff
@@ -69,27 +68,66 @@ public ModulePropertyAttribute()
         }
     }
 
+    internal static partial class TypeOffset
+    {
+        static TypeOffset()
+        {
+            Type type = typeof(TypeOffset);
+            FieldInfo[] fields = type.GetFields();
+            int size = IntPtr.Size;
+            for (int i = 0; i < fields.Length; i++)
+            {
+                int offset = i * size;
+                FieldInfo fi = fields[i];
+                fi.SetValue(null, offset);
+            }
+        }
+
+        public static int magic() => ManagedDataOffsets.Magic;
+    }
+
     internal static class ManagedDataOffsets
     {
-        static ManagedDataOffsets()
+        public static int Magic { get; private set; }
+        public static readonly Dictionary<string, int> NameMapping = new Dictionary<string, int>();
+
+        static class DataOffsets
         {
-            FieldInfo[] fi = typeof(ManagedDataOffsets).GetFields(BindingFlags.Static | BindingFlags.Public);
-            for (int i = 0; i < fi.Length; i++)
+            public static readonly int ob_data;
+            public static readonly int ob_dict;
+
+            static DataOffsets()
             {
-                fi[i].SetValue(null, -(i * IntPtr.Size) - IntPtr.Size);
+                FieldInfo[] fields = typeof(DataOffsets).GetFields(BindingFlags.Static | BindingFlags.Public);
+                for (int i = 0; i < fields.Length; i++)
+                {
+                    fields[i].SetValue(null, -(i * IntPtr.Size) - IntPtr.Size);
+                }
             }
+        }
 
-            size = fi.Length * IntPtr.Size;
+        static ManagedDataOffsets()
+        {
+            Type type = typeof(TypeOffset);
+            foreach (FieldInfo fi in type.GetFields())
+            {
+                NameMapping[fi.Name] = (int)fi.GetValue(null);
+            }
+            Magic = TypeOffset.members;
+            FieldInfo[] fields = typeof(DataOffsets).GetFields(BindingFlags.Static | BindingFlags.Public);
+            size = fields.Length * IntPtr.Size;
         }
 
-        public static readonly int ob_data;
-        public static readonly int ob_dict;
+        public static int GetSlotOffset(string name)
+        {
+            return NameMapping[name];
+        }
 
         private static int BaseOffset(IntPtr type)
         {
             Debug.Assert(type != IntPtr.Zero);
             int typeSize = Marshal.ReadInt32(type, TypeOffset.tp_basicsize);
-            Debug.Assert(typeSize > 0 && typeSize <= ExceptionOffset.Size());
+            Debug.Assert(typeSize > 0);
```

Comment:
@amos402 @koubaa 
As far as I can see, this is not actually fixing or explaining anything. It just removes the assert, that caused the problem.

Reply:
IMHO, the assert should stay. This PR does not address it, but probably solves the problem @koubaa has.
I have more thorough fix in my fork, but it is quite extensive, and this will do for now.

Reply:
The checking of `typeSize <= ExceptionOffset.Size()` is totally a wrong idea, there's no connection between them. Why an other object size whould relates to an exception object size.
You can simply make it lager than that size by
 ```python
class A(B, C, D, E, F, G): pass
```

Reply:
@amos402 if I remember correctly, `BaseOffset` is only used for reflections of .NET classes created by Python.NET. Right now there are only two kinds of them: `System.Exception` and everything derived from it, that end up deriving Python's `builtins.BaseException` or `builtins.Exception`, and every other .NET type, which end up deriving from `builtins.object`. Within those two categories each of the types has identical structure, and the ones derived from `Exception` just happen to be larger.

When Python class is inherited from a reflected .NET class via single inheritance, and has no extra low-level PyType stuff defined explicitly, I see no reason for its objects to be larger, than corresponding base class objects. Hence all such types should have either the same size as reflected `System.Object` or as reflected `System.Exception`.

Reply:
@lostmsu @amos402  if the check needs to be removed, it can be done in a follow-on PR.  I don't think this discussion should hold this PR.

Reply:
@koubaa  I am fine with this PR without the check removal.